src/mongoc/mongoc-write-command.c

```diff
@@ -973,12 +1015,12 @@ _mongoc_write_result_merge_legacy (mongoc_write_result_t  *result, /* IN */
           * upsert if _id is not an ObjectId.
           */
          result->nUpserted += 1;
-         if (bson_iter_init_find (&iter, command->u.update.update, "_id") ||
```

Comment:
pre-2.6 mongod didn't return an _id for an upsert if _id wasn't an oid.  So we have to go dig the upserted id out of the update or the selector.

Basically, the conditional breaks down as:

* n == 1
  * we ran a single update
* updatedExisting is boolean false
  * we didn't update, so an upsert happened

Then we check for an _id field in:

* the update
* the selector

The || short circuits, so we append whichever one we find first.

The whole thing should work because we only get there if n == 1 and upserted wasn't an oid, which means the server couldn't have generated one and we should have provided one.

It's a bit convoluted, but that's the logic